### ISSUE TITLE
Question: how to handle UI state disabled vs hidden

### ISSUE THREAD

**Issue by lifeiscontent** [palkan/action_policy#128]
Hey @palkan 

Hope you're doing well.

I've realized that sometimes I want to show/hide UI depending on an authorization, but how would you go about handling UI that can be disabled (with a tooltip e.g. message from authorization) and just hiding the UI all together? the `value` from authorization is not enough information to describe the state between disabled vs hidden

Have you dealt with this before?

**Comment by palkan** [comment]
I don't think that's the responsibility of a policy to decide on the UI state. 

You can try to incorporate [reasons](https://actionpolicy.evilmartians.io/#/reasons) for that or even reason [details](https://actionpolicy.evilmartians.io/#/reasons?id=detailed-reasons).

I used this approach in combination with GraphQL to display different UI elements depending on the reasons. We have a helper for that: `allowance_to` https://github.com/palkan/action_policy-graphql/blob/5e1772ac844fa67b1bf54539ed9178d59aa277eb/lib/action_policy/graphql/fields.rb#L29. It's an extended version of `allowed_to?` which return the whole result object, not only true or false. We should probably promote it to the main library.

**Comment by lifeiscontent** [comment]
@palkan let me break down my issue this is what I'm currently doing in JavaScript but I'd obviously like to move this to the backend.

```js
function canFavoriteOn(job) {
  let value = false;
  let disabled = false;
  let fullMessages = [];
  let reasons = { fullMessages };

  if (
    (job.jobApplication === null ||
      ![
        'accepted_offer',
        'considering_offer',
        'interviewing',
        'requested_interview',
        'withdrawn',
      ].includes(job.match.userState)) &&
    job.state === 'active'
  ) {
    value = true;
  }

  if (job.match.userState === 'interested') {
    disabled = true;
    fullMessages.push('You have expressed interest in this match');
  }

  return { value, disabled, reasons };
}
```

I know I've kinda bastardized the authorization object, but essentially what I'm saying here is as follows:

- if the jobApplication is not in one of `these` states, show it.
- if the match has already been favorited, disable it and show the reason why the user cannot perform the action.

**Comment by lifeiscontent** [comment]
I guess one of the ways I could do this is just do a check to show the button based off the jobApplication and then only allow the user to complete the action if the userState is not interested.

**Comment by palkan** [comment]
Let me try to rewrite the code above into a policy:

```ruby
class JobPolicy < ApplicationPolicy
  def favorite?
    check?(:interested?) && (
      no_application? ||
      active_state?
    )
  end

  def no_application?
    record.application.nil?
  end
  
  def active_state?
    user_state.in?(%w[accepted_offer ...])
  end

  def interested?
     details[:interested] =  (user_state == "interested")
  end

  private

  def user_state
    # ...
  end
end
```

Locales:

```yml
en:
  action_policy:
    policy:
      job:
        interested: 'You have expressed interest in this match'
```

Then, if you call:

```ruby
# somewhere in controller
allowance = policy_for(record: job).then do |policy|
  policy.apply(:favorite?)
  policy.result
end

render json: { value: result.value, fullMessages: result.reasons.full_messages, details: result.reasons.details }
```

And then you can check for `data.details["interested"]` in your JS code and use `data.fullMessages` for messages.

**Comment by lifeiscontent** [comment]
@palkan would this work for the GraphQL Authorization or only from a rails controller response?

**Comment by palkan** [comment]
That's a generic code for any _behaviour_; in action_policy-graphql we already have a similar functionality built-in.

**Comment by palkan** [comment]
@lifeiscontent `allowance_to` has been added to master and will be released in 0.5.0: https://github.com/palkan/action_policy/commit/3e749b225d0192ee159a76ae06d37b2ed1784ca0
